server/pulp/plugins/migration/standard_storage_path.py

```diff
@@ -0,0 +1,374 @@
+
+"""
+Provides common classes used to migrate content unit storage to
+the standard path format introduced in pulp 2.8. Each plugin needs
+to add a migration using these classes.
+For example:
+
+from pulp.plugins.migration.standard_storage_path import Migration, Plan
+from pulp.server.db import connection
+
+
+def migrate(*args, **kwargs):
+    key_fields = (.,..,.)
+    collection = connection.get_collection('units_rpm')
+    migration = Migration()
+    plan = Plan(collection, key_fields)
+    migration.add(plan)
+    migration()
+"""
+import os
+import shutil
+
+from collections import namedtuple
+from hashlib import sha256
+from itertools import chain
+
+from pulp.plugins.util.misc import mkdir
+from pulp.server.config import config
+
+
+# stored in the Batch
+Item = namedtuple(
+    'Item',
+    [
+        'plan',
+        'unit_id',
+        'storage_path',
+        'new_path'
+    ]
+)
+
+
+class Batch(object):
+    """
+    A batch of units to be migrated.
+
+    :ivar items: A dictionary of Item keyed by storage path.
+    :type items: dict
+    """
+
+    # The number of items in each batch.
+    # Using 500k based on memory profile of 100 mb / 100k.
+    LIMIT = 500000
+
+    def __init__(self):
+        self.items = {}
+
+    def add(self, unit):
+        """
+        Add the specified unit to the batch.
+
+        :param unit: A unit to add.
+        :type unit: Unit
+        """
+        item = Item(unit.plan, unit.id, unit.storage_path, unit.new_path)
+        self.items[unit.storage_path] = item
+
+    def reset(self):
+        """
+        Reset the batch by clearing added units.
+        """
+        self.items = {}
+
+    def _relink(self):
+        """
+        Update (relink) symlinks to reflect the storage path of migrated units.
+        Foreach symlink found, find the *new* path using the plan.  Then,
+        re-create the symlink with the updated target.
+        """
+        root = Migration.publish_dir()
+        for path, directories, files in os.walk(root):
+            for name in chain(files, directories):
```

Comment:
It's interesting that a symlink to a directory will show up in the ``directories`` list here. I wasn't sure which way it would go, but I did a quick test that confirms what you also concluded.